components/_impl/workers/subprocess_rpc.py

```diff
@@ -0,0 +1,488 @@
+
+    This class handles:
+        1) File descriptor lifetimes
+        2) File descriptor inheritance
+        3) Message packing and unpacking
+        4) (Optional) timeouts for reads
+    """
+
+    def __init__(
+        self,
+        read_handle: typing.Optional[int] = None,
+        write_handle: typing.Optional[int] = None,
+        timeout: typing.Optional[float] = None,
+        timeout_callback: typing.Callable[[], typing.NoReturn] = (lambda: None),
+    ) -> None:
+        self._owns_pipe = read_handle is None and write_handle is None
+        if self._owns_pipe:
+            self.read_fd, self.write_fd = os.pipe()
+        else:
+            self.read_fd = from_handle(read_handle, os.O_RDONLY)
+            self.write_fd = from_handle(write_handle, os.O_WRONLY)
+
+        self.read_handle = read_handle or to_handle(self.read_fd)
+        self.write_handle = write_handle or to_handle(self.write_fd)
+        self.timeout = timeout
+        self.timeout_callback = timeout_callback
+
+    def _read(self, size: int) -> bytes:
+        """Handle the low level details of reading from the PIPE."""
+        if self.read_fd is None:
+            raise IOError("Cannot read from PIPE, we do not have the read handle")
+
+        with _TimeoutPIPE.maybe_timeout_read(self):
+            raw_msg = os.read(self.read_fd, len(_CHECK) + size)
+
+        check_bytes, msg = raw_msg[:len(_CHECK)], raw_msg[len(_CHECK):]
+        if check_bytes == _TIMEOUT:
+            self.timeout_callback()  # Give caller the chance to cleanup.
+            raise IOError(f"Exceeded timeout: {self.timeout}")
+
+        if check_bytes != _CHECK:
+            raise IOError(f"{check} != {_CHECK}, {msg}")
+
+        if len(msg) != size:
+            raise IOError(f"len(msg) != size: {len(msg)} vs. {size}")
+
+        return msg
+
+    def read(self) -> bytes:
+        msg_size = struct.unpack(_ULL, self._read(_ULL_SIZE))[0]
+        return self._read(msg_size)
+
+    def write(self, msg: bytes) -> None:
+        if self.write_fd is None:
+            raise IOError("Cannot write from PIPE, we do not have the write handle")
+        assert isinstance(msg, bytes), msg
+        packed_msg = (
+            # First read: message length
+            _CHECK + struct.pack(_ULL, len(msg)) +
+
+            # Second read: message contents
+            _CHECK + msg
+        )
+
+        os.write(self.write_fd, packed_msg)
+
+    def _close_fds(self):
+        """Factor cleanup to a helper so we can test when it runs."""
+        os.close(self.read_fd)
+        os.close(self.write_fd)
+
+    def __del__(self) -> None:
+        if self._owns_pipe:
+            self._close_fds()
+
+
+# =============================================================================
+# == Exception Propagation  ===================================================
+# =============================================================================
+
+class ExceptionUnpickler(pickle.Unpickler):
```

Comment:
Great eye for detail. I would have just throwing "SubprocessException(Foo)" and made people unpack it.

Reply:
This is true dedication to transparency.

Reply:
Thanks! The rethrowing might be my favorite part of the entire PR.

Reply:
Thanks! The rethrowing might be my favorite part of the entire PR.